File: src/main/java/com/jcloisterzone/game/phase/CourierPhase.java
package com.jcloisterzone.game.phase;

import com.jcloisterzone.Player;
import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.board.pointer.MeeplePointer;
import com.jcloisterzone.event.ExprItem;
import com.jcloisterzone.event.CastleCreated;
import com.jcloisterzone.event.MeepleDeployed;
import com.jcloisterzone.event.PointsExpression;
import com.jcloisterzone.event.ScoreEvent;
import com.jcloisterzone.feature.Castle;
import com.jcloisterzone.feature.Completable;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.Monastery;
import com.jcloisterzone.feature.Scoreable;
import com.jcloisterzone.figure.Follower;
import com.jcloisterzone.figure.Meeple;
import com.jcloisterzone.figure.neutral.Courier;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.random.RandomGenerator;
import com.jcloisterzone.reducers.AddPoints;
import com.jcloisterzone.reducers.MoveNeutralFigure;

import io.vavr.Predicates;
import io.vavr.Tuple2;
import io.vavr.collection.Array;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;
import io.vavr.control.Option;

public class CourierPhase extends Phase {

    public CourierPhase(RandomGenerator random, Phase defaultNext) {
        super(random, defaultNext);
    }

    @Override
    public StepResult enter(GameState state) {
        FeaturePointer fpCourier = state.getNeutralFigures().getCourierDeployment();
        if (fpCourier == null || fpCourier.getPosition() == null) {
            return next(state);
        }

        GameState _state = state;

        Set<Location> restrictedQuarters = HashSet.of(
      	    Location.QUARTER_CASTLE, 
       	    Location.QUARTER_MARKET, 
       	    Location.QUARTER_BLACKSMITH, 
       	    Location.QUARTER_CATHEDRAL
       	);
        
        List<MeepleDeployed> deployedThisTurn = List.ofAll(state.getCurrentTurnPartEvents())
       	    .filter(Predicates.instanceOf(MeepleDeployed.class))
       	    .map(ev -> (MeepleDeployed) ev)
       	    .filter(ev -> (ev.getMeeple() instanceof Follower))
       	    .filter(ev -> !restrictedQuarters.contains(ev.getLocation()));
        
        boolean castleCreatedThisTurn = List.ofAll(state.getCurrentTurnPartEvents())
            .exists(Predicates.instanceOf(CastleCreated.class));

        // Meeples moved to a Castle this turn and still deployed there
        List<MeepleDeployed> castleEvents = castleCreatedThisTurn
            ? deployedThisTurn.filter(ev ->
                ev.getMovedFrom() != null
                && ((FeaturePointer) ev.getPointer()).getFeature().equals(Castle.class)
                && _state.getDeployedMeeples().get(ev.getMeeple())
                         .map(ev.getPointer()::equals).getOrElse(false))
            : List.empty();

        MeepleDeployed meepleForCourierMessage = deployedThisTurn
            .flatMap(ev -> {
                Meeple m = ev.getMeeple();
                FeaturePointer currentFp = _state.getDeployedMeeples().get(m).getOrNull();
                if (currentFp == null) return List.empty();
                if (!meepleOrthogonallyAlignedToCourier(_state, new Tuple2<>(m, currentFp), fpCourier)) return List.empty();
                // Still at original deployment position
                if (currentFp.equals(ev.getPointer())) return List.of(ev);
                // Promoted to castle lord this turn
                return castleEvents.filter(mev -> mev.getMeeple().equals(m)).take(1);
            })
            .headOption()
            .getOrNull();

        if (meepleForCourierMessage == null) {
            return next(state);
        }

        state = new MoveNeutralFigure<>(
            state.getNeutralFigures().getCourier(),
            meepleForCourierMessage.getPointer()
        ).apply(state);

        Position pos = meepleForCourierMessage.getPointer().getPosition();
        Feature courierFeature = state.getFeature((FeaturePointer) meepleForCourierMessage.getPointer());

        List<Tuple2<Meeple, FeaturePointer>> affectedMeeples = state.getDeployedMeeples()
            .filter((m, fp) -> {
                if (!(m instanceof Follower)) return false;
                Feature f = _state.getFeature(fp);
                if (!(f instanceof Completable)) return false;
                if (f instanceof Monastery && ((Monastery) f).isSpecialMonastery(_state)) return false;
                Position mpos = fp.getPosition();
                if (courierFeature instanceof Castle) {
                    return ((Castle) courierFeature).getVicinity().contains(mpos);
                }
                return Math.abs(pos.x - mpos.x) <= 1 && Math.abs(pos.y - mpos.y) <= 1;
            })
            .toList()
            .distinctBy(t -> _state.getFeature(t._2()));

        for (Tuple2<Meeple, FeaturePointer> affected : affectedMeeples) {
            Completable feature = (Completable) state.getFeature(affected._2);
            PointsExpression expr = feature.getPoints(state);
            ScoreEvent.ReceivedPoints points = new ScoreEvent.ReceivedPoints(
                new PointsExpression("courier." + expr.getName(), expr.getItems()),
                meepleForCourierMessage.getMeeple().getPlayer(),
                affected._1.getDeployment(state)
            );
            state = new AddPoints(points, false).apply(state);
        }

        return next(state);
    }

    private Set<Position> featurePositions(GameState state, FeaturePointer fp) {
        Feature feature = state.getFeature(fp);
        if (feature instanceof Castle) {
            return HashSet.ofAll(((Castle) feature).getPlaces().map(FeaturePointer::getPosition));
        }
        return HashSet.of(fp.getPosition());
    }

    boolean meepleOrthogonallyAlignedToCourier(
            GameState state,
            Tuple2<Meeple, FeaturePointer> meeple,
            FeaturePointer courier) {
        Set<Position> meeplePos  = featurePositions(state, meeple._2);
        Set<Position> courierPos = featurePositions(state, courier);
        return meeplePos.exists(mp -> courierPos.exists(cp -> mp.x == cp.x || mp.y == cp.y));
    }
}

File: src/main/ts/com/jcloisterzone/game/phase/CourierPhase.ts
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import type { BoardPointer } from "../../board/pointer/BoardPointer.js";
import type { FeaturePointer } from "../../board/pointer/FeaturePointer.js";
import { Position } from "../../board/Position.js";
import { Location } from "../../board/Location.js";
import { CastleCreated } from "../../event/CastleCreated.js";
import { MeepleDeployed } from "../../event/MeepleDeployed.js";
import { PointsExpression } from "../../event/PointsExpression.js";
import { ReceivedPoints } from "../../event/ScoreEvent.js";
import { Castle } from "../../feature/Castle.js";
import { City } from "../../feature/City.js";
import { type Completable, isInstanceOfCompletable } from "../../feature/Completable.js";
import type { Feature } from "../../feature/Feature.js";
import { Garden } from "../../feature/Garden.js";
import { Monastery } from "../../feature/Monastery.js";
import { River } from "../../feature/River.js";
import { Road } from "../../feature/Road.js";
import { YagaHut } from "../../feature/YagaHut.js";
import { Follower } from "../../figure/Follower.js";
import type { Meeple } from "../../figure/Meeple.js";
import type { NeutralFigure } from "../../figure/neutral/NeutralFigure.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { AddPoints } from "../../reducers/AddPoints.js";
import { MoveNeutralFigure } from "../../reducers/MoveNeutralFigure.js";
import type { GameState } from "../state/GameState.js";
import { Phase } from "./Phase.js";
import type { StepResult } from "./StepResult.js";


/** The Courier walks to a follower deployed in its row/column this turn and scores that
 *  follower's neighbouring completed features for the courier owner. */
export class CourierPhase extends Phase {
  static readonly simpleName = "CourierPhase";

  constructor(random: RandomGenerator, defaultNext: Phase | null) {
    super(random, defaultNext);
  }

  enter(state: GameState): StepResult {
    const fpCourier = state.getNeutralFigures().getCourierDeployment();
    if (fpCourier === null || fpCourier.getPosition() === null) {
      return this.next(state);
    }

    const restrictedQuarters = [
      Location.QUARTER_CASTLE,
      Location.QUARTER_MARKET,
      Location.QUARTER_BLACKSMITH,
      Location.QUARTER_CATHEDRAL,
    ];
    const deployedThisTurn: MeepleDeployed[] = [];
    let castleCreatedThisTurn = false;
    for (const ev of state.getCurrentTurnPartEvents()) {
      if (ev instanceof CastleCreated) castleCreatedThisTurn = true;
      if (
        ev instanceof MeepleDeployed &&
        ev.getMeeple() instanceof Follower &&
        !restrictedQuarters.includes(ev.getLocation()!)
      )
        deployedThisTurn.push(ev);
    }

    // meeples that moved into a castle this turn and are still its lord — keep the
    // RE-EMITTED castle event (its pointer is the Castle, not the original City)
    const castleEvents: MeepleDeployed[] = castleCreatedThisTurn
      ? deployedThisTurn.filter((ev) => {
          const cur = state.getDeployedMeeples().get(ev.getMeeple()).getOrNull();
          return (
            ev.getMovedFrom() !== null &&
            ev.getPointer().asFeaturePointer().getFeature() === (Castle as never) &&
            cur !== null &&
            cur.equals(ev.getPointer().asFeaturePointer())
          );
        })
      : [];

    let meepleForCourier: MeepleDeployed | null = null;
    for (const ev of deployedThisTurn) {
      const m = ev.getMeeple();
      const currentFp = state.getDeployedMeeples().get(m).getOrNull();
      if (currentFp === null) continue;
      if (!this.meepleOrthogonallyAlignedToCourier(state, currentFp, fpCourier)) continue;
      if (currentFp.equals(ev.getPointer().asFeaturePointer())) {
        meepleForCourier = ev;
        break;
      }
      // promoted to castle lord this turn → use the castle event (Castle pointer)
      const castleEv = castleEvents.find((cev) => cev.getMeeple().equals(m));
      if (castleEv !== undefined) {
        meepleForCourier = castleEv;
        break;
      }
    }

    if (meepleForCourier === null) {
      return this.next(state);
    }

    const targetFp = meepleForCourier.getPointer().asFeaturePointer();
    state = new MoveNeutralFigure(
      state.getNeutralFigures().getCourier() as unknown as NeutralFigure<BoardPointer>,
      meepleForCourier.getPointer(),
    ).apply(state);

    const pos = targetFp.getPosition();
    const courierFeature = state.getFeature(targetFp);
    const owner = meepleForCourier.getMeeple().getPlayer();

    // affected: occupied completables near the courier's new feature (deduped by feature)
    const affected: Array<[Meeple, FeaturePointer]> = [];
    const seenFeatures = new globalThis.Set<Feature>();
    for (const t of state.getDeployedMeeples()) {
      const m = t._1;
      const fp = t._2;
      if (!(m instanceof Follower)) continue;
      const f = state.getFeature(fp);
      if (f === null || !isInstanceOfCompletable(f)) continue;
      if (f instanceof Monastery && f.isSpecialMonastery(state)) continue;
      const mpos = fp.getPosition();
      let inRange: boolean;
      if (courierFeature instanceof Castle) {
        inRange = courierFeature.getVicinity().contains(mpos);
      } else {
        inRange = Math.abs(pos.x - mpos.x) <= 1 && Math.abs(pos.y - mpos.y) <= 1;
      }
      if (!inRange) continue;
      if (seenFeatures.has(f)) continue;
      seenFeatures.add(f);
      affected.push([m, fp]);
    }

    for (const [m, fp] of affected) {
      const feature = state.getFeature(fp) as unknown as Completable;
      const expr = feature.getPoints(state);
      const points = new ReceivedPoints(
        new PointsExpression("courier." + expr.getName(), expr.getItems()),
        owner,
        m.getDeployment(state),
      );
      state = new AddPoints(points, false).apply(state);
    }

    return this.next(state);
  }

  private featurePositions(state: GameState, fp: FeaturePointer): Set<Position> {
    const feature = state.getFeature(fp);
    if (feature instanceof Castle) {
      let s: Set<Position> = HashSet.empty<Position>();
      for (const p of feature.getPlaces()) s = s.add(p.getPosition());
      return s;
    }
    return HashSet.of(fp.getPosition());
  }

  private meepleOrthogonallyAlignedToCourier(
    state: GameState,
    meepleFp: FeaturePointer,
    courierFp: FeaturePointer,
  ): boolean {
    const meeplePos = this.featurePositions(state, meepleFp);
    const courierPos = this.featurePositions(state, courierFp);
    return meeplePos.exists((mp) => courierPos.exists((cp) => mp.x === cp.x || mp.y === cp.y));
  }
}
